### ISSUE TITLE
[no-unused-vars] False positives on required args in interface implementations

### ISSUE THREAD

**Issue by RingOfStorms** [typescript-eslint/typescript-eslint#944]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": ["error"]
  }
}
```

```TS
class A {
  abstract doAFunkyDance(withStyle: boolean): void;
}

class B extends A {
  doAFunkyDance(withStyle: boolean) { // Error here withStyle not used
     // Class B always dances with style so we don't end up using the var... but in typescript, it MUST be here due to the interface of class A
  }
}

class C extends A {
  doAFunkyDance(withStyle: boolean) { // Error here withStyle not used
    throw new Error('Class C does not knwo how to do a funky dance');
  }
}

class Z extends A {
  doAFunkyDance(withStyle: boolean) {
    if (withStyle) { ... } else { ... } // Valid use, no error in the class
  }
}
```

**Expected Result**
When writing an interface with params all of its implementations must declare those params in their implementation regardless of use. I believe that `no-unused-vars` should not error out when a function is an implementation of a defined interface. There are valid use cases where at least one implementation may not use all the parameters, but it cannot be removed from the declaration without breaking the type checking of consumers.

**Actual Result**
no-unused-vars throws lint errors on the lines indicated in the example.

**Additional Info**
NA

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.13.0` |
| `@typescript-eslint/parser`        | `1.13.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `v8.12.0` |
| `npm`                              | `5.7.1` |

**Comment by bradzacher** [comment]
First of all, let me preface this by saying this is an easy fix - use the `argsIgnorePattern` option. The common thing to do is to prefix args with an underscore if they are intentionally unused.

----

Secondly, we can't automatically catch this case. TypeScript itself doesn't even catch this case in its `noUnusedParameters` implementation.

Technically it's valid to omit the parameter - you can omit arguments from an abstract/interface method, but you cannot add non-optional arguments.

When casting to the interface, it works fine.

i.e.

```
class A {
  abstract doAFunkyDance(withStyle: boolean): void;
}

class B extends A {
  doAFunkyDance() {}
}

class B extends A {
  doAFunkyDance(withStyle: boolean) {}
}

const instance: A = new B();


```
